### ISSUE TITLE
Research/enhance audio formats/codecs (Opus/FLAC) #Idea #Suggestion

### ISSUE THREAD

**Issue by ldexterldesign** [mastodon/mastodon#17037]
### Pitch

Hi 👋,

I hope y'all well

Firstly, thanks for superb software

Secondly, I'm depressed Opus is still unsupported

I see we support MP3, which is technically open (patent, DRM) now according to [comparison], but I researched and discovered Opus (lossy, newer than Vorbis+OGG) and FLAC (lossless) are arguably the best (open) codec and format, respectively.

Given Mastodon's ethos shouldn't we be aiming to use these open technologies and gracefully degrade for devices/platforms/clients who don't support them..?

@Gargron, some 2.5 years later, is [this] still your position?:
- If no then perhaps we can revisit this issue here
- If yes then no hard feelings closing this issue; perhaps I'll wait another 2.5 years

If you're reading and:
- agree, then please vote up (e.g. +1, like), or;
- disagree, then please vote down (e.g. -1, dislike) - I'd love to know why?

Welcome feedback/input

Hope to hear back

Sincerely

[comparison]: https://en.wikipedia.org/wiki/Comparison_of_audio_coding_formats
[this]: https://github.com/mastodon/mastodon/issues/11131

### Motivation

I researched what's a modern audio codec/format and it's Opus

**Comment by trwnh** [comment]
given the situation still hasn't meaningfully improved (*shakes fist at safari*), i'd say we're stuck with mp3 for a while longer...

https://caniuse.com/?search=opus -- safari on both macos and ios technically supports opus, but only when packaged in their Core Audio Format
https://caniuse.com/?search=ogg -- safari on ios still doesn't support ogg vorbis at all

**Comment by ldexterldesign** [comment]
Thanks for reply but should we not care what Safari does - isn’t that the point of progressive enhancement and graceful degradation?

Cheers

**Comment by Gargron** [comment]
Yes, two files instead of one. If the goal is to decrease file size, then that defeats the point.

**Comment by ldexterldesign** [comment]
Hi @Gargron,

Thanks for reply

Mmm, I can understand why a server owner wouldn't want to incur these (extra) costs right now

--

🤔

Using FFMPEG standard settings I output two identical FLAC files:
1. MP3, 3.7MB
2. OPUS, 2.6MB

Based on this, OPUS 70% size of MP3 so a server owner could be saving 30% disk space right now

Regards

**Comment by trwnh** [comment]
@ldexterldesign opus uses less space than mp3 but is not supported on safari outside of apple's Core Audio Format as mentioned above. it makes no sense to store both mp3 and opus, because that would necessarily take more space than using only mp3.

**Comment by MrAureliusR** [comment]
> @ldexterldesign opus uses less space than mp3 but is not supported on safari outside of apple's Core Audio Format as mentioned above. it makes no sense to store both mp3 and opus, because that would necessarily take more space than using only mp3.

Or we could, again, not let Apple's closed ecosystem dictate what an open-source project does? Do Mac users seriously still use the default browser? This sounds like a Safari problem, not a mastodon problem.

**Comment by trwnh** [comment]
> Do Mac users seriously still use the default browser?

iOS *only* has Safari. iOS powers about 30% of mobile devices. Using *only* a format that excludes 30% of mobile users is not good. Neither is using *two* formats instead of just one.

**Comment by MrAureliusR** [comment]
You can't install Chrome or Firefox on iOS? And even if that's the case, why let Apple hold Mastodon back? I bet if you looked at the people who actually use Mastodon, there are likely very, very few users accessing it via the mobile browser on iOS devices. There are plenty of great Mastodon apps on Android and I'm presuming there's at least one or two decent ones on Apple. This is how the vast majority of users access social media on mobile platforms anyway -- people use the Twitter/Instagram/Facebook apps instead of visiting the mobile browser page.

And again, I strongly believe that Mastodon should be supporting open standards like Opus, Theora and FLAC and help push the platforms into supporting this rather than crippling Mastodon to fit in with them.

**Comment by ldexterldesign** [comment]
I use Mastodon on iOS Safari because I don’t like installing proprietary apps and want to support the open web

If I could install my browser of choice in iOS I would but that’s on Apple and will probably never change so based on this I created this issue

**Comment by ldexterldesign** [comment]
If we’re letting data analytics prevail then we should be optimising for the majority; 70% of mobile users and providing fallbacks for the 30%. I don’t see it as a waste of work.

Ask server owners if they care more about disk space / CPU (no modern file support / transcode, single file) or bandwidth (modern file support, multiple files)

**Comment by brandonwkipp** [comment]
> iOS _only_ has Safari.

This is not true; I literally downloaded Firefox on iOS a few days ago and already have Chrome and Opera installed as well.

**Comment by trwnh** [comment]
they're still using safari's webkit engine on the backend

**Comment by ldexterldesign** [comment]
> Due to the addition of WebRTC support in Apple's WebKit rendering engine, macOS High Sierra and iOS 11 come with native playback support for Opus audio encapsulated in Core Audio Format containers. - https://en.wikipedia.org/wiki/Opus_(audio_format)#Operating_system_support

**Comment by ldexterldesign** [comment]
I don't know enough about it [WebRTC] but could it be possible to implement this for iOS using it?

**Comment by trwnh** [comment]
CAF is a special container format that only works for Apple. You still
can't play Opus outside of CAF or streaming situations (WebRTC is like
having a realtime connection for video calls or audio calls; in this
situation Opus would be delimited by the network packets rather than any
container format). This would mean having to store two files -- a regular
Opus file in a standard container (like OGG, WebM, etc), and an
Apple-specific Opus file in a CAF container.

**Comment by trwnh** [comment]
> Opus will never get Apple support, which means neither will Mastodon

well, there's a few ways out of this:

1) apple finally decides to support opus natively at some point in the future (hey, it *could* happen)
2) mastodon implements some javascript decoder such as ogv.js (i'm not sure how this affects native apps which are not browsers)
3) mastodon decides the disadvantage of storing two files is worth the advantage in less bandwidth for most users (although i can't say bandwidth is a larger concern than storage space for most admins...)

**Comment by ldexterldesign** [comment]
> Ask server owners if they care more about disk space / CPU (no modern file support / transcode, single file) or bandwidth (modern file support, multiple files). You’ll also see (I bet) how most server owners will want to support open / modern standards.

**Comment by RinCat** [comment]
Any update?

**Comment by TheEssem** [comment]
It appears that as of version 18.4, released in March, [Safari now supports the Ogg container](https://developer.apple.com/documentation/safari-release-notes/safari-18_4-release-notes#Media).

Given that a large amount of users are still on older versions, however, I feel like AAC-LC would be an alright replacement target audio codec - support is widespread, it's higher quality than MP3 with lower filesize, and the patents expired many years ago. There are a couple of potential caveats, however:
- Firefox support depends on the installed system codecs (though this could prove to be a non-issue in practice)
- Quality is heavily dependent on the encoder; FFmpeg's built-in encoder is considered to be lower quality than many other options, and the license of the highest-quality source-available/cross-platform encoder (FDK-AAC) is slightly controversial
